java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,943 @@
+import java
+private import RegexFlowConfigs
+
+/**
+ * A string literal that is used as a regular exprssion.
+ */
+abstract class RegexString extends Expr {
+  RegexString() { this instanceof StringLiteral }
+
+  /** Holds if a character set starts between `start` and `end`. */
+  private predicate char_set_start0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (
+      this.getChar(start + 1) = "^" and end = start + 2
+      or
+      not this.getChar(start + 1) = "^" and end = start + 1
+    )
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate char_set_end0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
+    not this.char_set_start0(_, pos - 1)
+  }
+
+  /**
+   * Gets the nesting depth of charcter classes at position `pos`
+   */
+  private int char_set_depth(int pos) {
+    exists(this.getChar(pos)) and
+    result =
+      max(int j |
+        j = 0 or
+        j =
+          count(int i | i < pos and this.char_set_start0(i, _)) -
+            count(int i | i < pos and this.char_set_end0(i))
+      )
+  }
+
+  /** Hold if a top-level character set starts between `start` and `end`. */
+  predicate char_set_start(int start, int end) {
+    this.char_set_start0(start, end) and
+    this.char_set_depth(start) = 0
+  }
+
+  /** Holds if a top-level character set ends at `pos`. */
+  predicate char_set_end(int pos) {
+    this.char_set_end0(pos) and
+    this.char_set_depth(pos) = 1
+  }
+
+  /**
+   * Whether there is a top-level character class, between start (inclusive) and end (exclusive)
+   *
+   * For now, nested character classes are approximated by only considering the top-level class for parsing.
+   * This leads to very similar results for ReDoS queries.
+   */
+  predicate charSet(int start, int end) {
+    exists(int inner_start, int inner_end |
+      this.char_set_start(start, inner_start) and
+      not this.char_set_start(_, start)
+    |
+      end = inner_end + 1 and
+      inner_end > inner_start and
+      this.char_set_end(inner_end) and
+      not exists(int mid | this.char_set_end(mid) | mid > inner_start and mid < inner_end)
```

Comment:
Would this be better expressed as a regex match of `[^]]+` or similar?

Reply:
I think that would require taking a substring, which is something that it looks like this library is trying to avoid except at the top-level by mostly handling indicies. Doing so could cause quadratically many substrings to be constructed.

Reply:
(additionally, that regex seems less clear to me what it's doing than this ql does. It also qould fail on escaped `]`s in the class, or nested classes)